### PR TITLE
feat: Embed - Introduce `prerender` instruction - Lightning fast popup experience

### LINKED ISSUE #11269: [CAL-2433] Embed: Allow preloaded iframe to be prefilled when actually showing the embed
Add a `prerenderIframe` option to `preload` instruction that would allow it the iframe to be reused fully and when the embed is actually supposed to be shown it would just toggle the preloaded iframe after applying the config provided.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2433](https://linear.app/calcom/issue/CAL-2433/embed-allow-preloaded-iframe-to-be-prefilled-when-actually-showing-the)</sub>

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

Fixes #11269 by introducing a `prerender` instruction that allows popups to load lightning fast. It works by loading the iframe(not visible) in the place where the modal would open. When the user clicks the popup CTA, it just reuses that hidden iframe and applies all configurations before showing it to user.

If you know beforehand that on particular page user is going to open a particular calLink popup, it can be prerendered.
- Added `options` and `type` to `preload` instruction.
```Javascript

Cal("prerender", {
    calLink: "free/30min",
    type: "modal" | "floatingButton",  // inline embed doesn't yet support prerendering, I think it won't be needed. When there is a usecase we can add that.
  });
```
- Fixes a bug where an harmless error (accessing `getBoundingClientRect` on undefined) was thrown for popup type embed
- A small refactor to have all embed fixtures under `embeds`. Plus, added a new fixture `gotoPlayground` that avoids the possibility of doing `page.goto `before adding embed listeners. That would have caused tests to give a false alarm.

[Demo - Prerendered Popup Embed](https://www.loom.com/share/Prerendered-popup-embed-7238da3f14ef4e4c9963003dd2efcd31)
[Demo - Regular popup Embed](https://www.loom.com/share/Non-prerendered-popup-embed-3edc1157c47c486f860533a81a5eed6f)

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?
- Click on 'Goto prerender test page only' in the embed playground. That would execute the preload API https://github.com/calcom/cal.com/pull/11303/files#diff-b82c210cc588aab8c102cf275cbcef37ca76f9ddf88da49f5e4407187d0c4e09R220
- After that clicking on 'Book with Free User[Dark Theme]' would show the popup almost instantly(with an intentional delay of 700ms)

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

File: apps/web/playwright/fixtures/embeds.ts
Comment (hariombalhara) [inline apps/web/playwright/fixtures/embeds.ts:1]: Changes to move individual embed related fixtures into **embeds** fixture

File: packages/embeds/embed-core/index.html [hunk 1]
```diff
@@ -85,20 +85,30 @@
...
           <i
-            >Corresponding Cal Link is being preloaded. Assuming that it would take you some time to click this
+            >Corresponding Cal Link is being prerendered. Assuming that it would take you some time to click this
             as you are reading this text, it would open up super fast[If you are running a production build on
             local]. Try switching to slow 3G or create a custom Network configuration which is impossibly
-            slow</i
+            slow. This should be used if you know beforehand which type of embed is going to be opened.</i
+          >`);
+        }
+        if (only === "all" || only === "preload-test") {
```
Comment (hariombalhara) [inline packages/embeds/embed-core/index.html:107]: Two different testcases for preloading and prerendering.

File: packages/embeds/embed-core/src/ModalBox/ModalBox.ts [hunk 2]
```diff
@@ -68,17 +83,23 @@ export class ModalBox extends HTMLElement {
       return;
     }
 
-    if (newValue == "loaded") {
-      this.getLoaderElement().style.display = "none";
-    } else if (newValue === "started") {
+    if (newValue === "loading") {
```
Comment (hariombalhara) [inline packages/embeds/embed-core/src/ModalBox/ModalBox.ts:86]: `started` -> `loading` better name for lifecycle state.

File: packages/embeds/embed-core/src/ModalBox/ModalBox.ts [hunk 3]
```diff
@@ -68,17 +83,23 @@ export class ModalBox extends HTMLElement {
...
+      this.open();
+      this.hideIframe();
+      this.getLoaderElement().style.display = "block";
+    } else if (newValue == "loaded" || newValue === "reopening") {
```
Comment (hariombalhara) [inline packages/embeds/embed-core/src/ModalBox/ModalBox.ts:90]: Added `reopening` state that is used when a popup CTA is clicked again, after it has already opened the embed.

File: packages/embeds/embed-core/src/ModalBox/ModalBox.ts [hunk 4]
```diff
@@ -68,17 +83,23 @@ export class ModalBox extends HTMLElement {
...
+      this.showIframe();
+      this.getLoaderElement().style.display = "none";
     } else if (newValue == "closed") {
       this.close();
     } else if (newValue === "failed") {
       this.getLoaderElement().style.display = "none";
       this.getErrorElement().style.display = "inline-block";
       const errorString = getErrorString(this.dataset.errorCode);
       this.getErrorElement().innerText = errorString;
+    } else if (newValue === "prerendering") {
```
Comment (hariombalhara) [inline packages/embeds/embed-core/src/ModalBox/ModalBox.ts:101]: `prerendering` state that doesn't show the embed and does things hidden from sight.

File: packages/embeds/embed-core/src/embed.ts [hunk 5]
```diff
@@ -398,6 +383,10 @@ export class Cal {
     });
 
     this.actionManager.on("linkReady", () => {
+      if (this.isPerendering) {
+        // Absolute check to ensure that we don't mark embed as loaded if it's prerendering otherwise prerendered embed would showup without any user action
```
Comment (hariombalhara) [inline packages/embeds/embed-core/src/embed.ts:387]: E2E also verifies this that embed should never be shown in prerending state

File: packages/embeds/embed-core/src/embed.ts [hunk 6]
```diff
@@ -563,49 +554,79 @@ class CalApi {
   modal({
     calLink,
     config = {},
-    uid,
+    __prerender = false,
```
Comment (hariombalhara) [inline packages/embeds/embed-core/src/embed.ts:557]: Not to be used by API users. Internal param.

File: packages/embeds/embed-core/src/embed.ts [hunk 7]
```diff
@@ -779,7 +849,6 @@ document.addEventListener("click", (e) => {
   api("modal", {
     calLink: path,
     config,
-    uid: modalUniqueId,
```
Comment (hariombalhara) [inline packages/embeds/embed-core/src/embed.ts:782]: `modal` would now compute its Id itself and store on Cal instance instead of storing on DOM element

**Review by zomars — COMMENTED** [review]
Awesome work @hariombalhara Code LGTM. Can we fix the conflicts before merging?

**Comment by hariombalhara** [comment]
Thanks @zomars 🙏 Resolved now.

**Comment by PeerRich** [comment]
wow this is so much faster!!!

**Review by PeerRich — APPROVED** [review]
approved on behalf of @zomars since he reviewed and approved this before merge conflicts

**Review by zomars — APPROVED** [review]
Ship it

**Pull request merged**